asmcomp/arm64/emit.mlp

```diff
@@ -789,8 +805,22 @@ let emit_instr env i =
         | Single ->
             `	fcvt	s7, {emit_reg src}\n`;
             `	str	s7, {emit_addressing addr base}\n`;
-        | Word_int | Word_val | Double ->
+        | Word_int | Double ->
             `	str	{emit_reg src}, {emit_addressing addr base}\n`
+        | Word_val ->
```

Comment:
If the `size` is `Word_val` then I believe we have the guarantee that the store will be an initialising store. Hence, `str` should be sufficient. @xavierleroy mentions the same here: https://github.com/ocaml/ocaml/pull/10972#issuecomment-1032704351. Perhaps add an assertion: `assert (not (size = Word_val && is_assignment))`.

Reply:
FWIW I've created https://github.com/ocaml/ocaml/pull/10999 to avoid conflating `Word_val` and `Word_int` in Cmm and Mach. https://github.com/ocaml/ocaml/pull/10999 does not block this PR.

Reply:
My comment is different: I though we agreed that non-initializing stores (of the "int" or "val" kind) should be `dmb ishld; str` (normal store) and not `dmb ishld; stlr` (release store), as no object publication can take place.

Reply:
I'm confused, [the table](https://github.com/ocaml/ocaml/pull/10972#issuecomment-1030596278) says `stlr` for non-initializing writes of pointers. Cannot `Word_val`-typed values contain pointers and thus require a `stlr` for publication safety? (The pointer write, into a shared location, may "publish" a block that was only known to the current domain before.)

Reply:
No, if the value can be a pointer, a call to `caml_modify` would have been generated instead of a direct store.  So, the value being stored is an integer and the value being overwritten is an integer.

Reply:
And the " `stlr` for non-initializing writes of pointers " in the table is for `caml_modify`.

Reply:
> No, if the value can be a pointer, a call to caml_modify would have been generated instead of a direct store.

Thanks, I missed that. For other people curious about this, `caml_modify` is generated at the following point:

https://github.com/ocaml/ocaml/blob/9093386715688b5055b3b3258da73bfca2433613/asmcomp/cmm_helpers.ml#L2199-L2210

Reply:
Indeed. `dmb ishld; str` is sufficient here. Sorry I wasn't clear. 

I was surprised to see `Istore(Word_val, _, true)` which should never appear as all such instances should go through `caml_modify`.

Reply:
Yes, there's something confusing in the `Word_val`/`Word_int` distinction.  It could be argued that stores of the `Word_int` kind are assignments to big arrays of 64-bit integers, and thus need no barrier, while `Word_val` assignments store tagged integers in OCaml references and other mutable data structures, and thus need a barrier...  And that's no the interpretation you had in mind :-)

Reply:
Indeed. I was confused about the interpretation of `Word_int`. Thanks for the clarification. I closed the misinformed PR #10999.

Reply:
No need to feel bad about it: the more I think about `Word_int` the more I find it confusing :-)  Until now, the `Word_int` / `Word_val` distinction mattered  only for loads, but not for stores.  We can introduce different behaviors for stores, but we need to agree on which behaviors are the most useful.  (And then document to death, of course.)  Let's keep this topic for a future PR (perhaps).